src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutorKeepalived.java

```diff
@@ -0,0 +1,320 @@
+      this.retrier = retrier;
+      this.callCredentialsProvider = callCredentialsProvider;
+      this.executeBackoff = this.retrier.newBackoff();
+      this.waitExecutionBackoff = new ProgressiveBackoff(this.retrier::newBackoff);
+      this.executionBlockingStubSupplier = executionBlockingStubSupplier;
+    }
+
+    ExecuteResponse start() throws IOException, InterruptedException {
+      // Execute has two components: the Execute call and (optionally) the WaitExecution call.
+      // This is the simple flow without any errors:
+      //
+      // - A call to Execute returns streamed updates on an Operation object.
+      // - We wait until the Operation is finished.
+      //
+      // Error possibilities:
+      // - An Execute call may fail with a retriable error (raise a StatusRuntimeException).
+      //   - If the failure occurred before the first Operation is returned and tells us the
+      //     execution is accepted, we retry the call.
+      //   - Otherwise, we call WaitExecution on the Operation.
+      // - A WaitExecution call may fail with a retriable error (raise a StatusRuntimeException).
+      //   In that case, we retry the WaitExecution call on the same operation object.
+      // - A WaitExecution call may fail with a NOT_FOUND error (raise a StatusRuntimeException).
+      //   That means the Operation was lost on the server, and we will retry to Execute.
+      // - Any call can return an Operation object with an error status in the result. Such
+      //   Operations are completed and failed; however, some of these errors may be retriable.
+      //   These errors should trigger a retry of the Execute call, resulting in a new Operation.
+      Preconditions.checkState(lastOperation == null);
+
+      ExecuteResponse response = null;
+      // Exit the loop as long as we get a response from either Execute() or WaitExecution().
+      while (response == null) {
+        // We use refreshIfUnauthenticated inside retry block. If use it outside, retrier will stop
+        // retrying when received a unauthenticated error, and propagate to refreshIfUnauthenticated
+        // which will then call retrier again. It will reset the retry time counter so we could
+        // retry more than --remote_retry times which is not expected.
+        response = retrier
+            .execute(() -> Utils.refreshIfUnauthenticated(this::execute, callCredentialsProvider),
+                executeBackoff);
+
+        // If no response from Execute(), use WaitExecution() in a "loop" which is implemented
+        // inside the retry block.
+        //
+        // The cases to exit the loop:
+        //   1. Received the final response.
+        //   2. Received a un-retriable gRPC error.
+        //   3. Received NOT_FOUND error where we will retry Execute() (by returning null).
+        //   4. Received consecutive retriable gRPC errors (up to max retry times).
+        if (response == null) {
+          response = retrier.execute(
```

Comment:
Do I understand correctly that if the WaitExecution() backoff runs out, we'll raise an exception? 

The scenario I'm looking to avoid is that we first wait for the WaitExecution() backoff to run out, then that results in bumping the Execute() backoff, and then this would signal an error after a N*M tries when one backoff running out causes the other backoff just to be bumped by one)

Reply:
Yes. It will rase an exception.

See test case `executeRemotely_executeAndRetryWait_failForConsecutiveErrors`.